presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java

```diff
@@ -5741,4 +5741,34 @@ public void testMapUnionSumOverflow()
                 "select y, map_union_sum(x) from (select 1 y, map(array['x', 'z', 'y'], cast(array[null,30, 32760] as array<smallint>)) x " +
                         "union all select 1 y, map(array['x', 'y'], cast(array[1,100] as array<smallint>))x) group by y", ".*Value 32860 exceeds MAX_SHORT.*");
     }
+
+    @Test
+    public void testLosslessJoin()
+    {
+        MaterializedResult countPart = computeActual("select partkey from part group by partkey");
+        MaterializedResult countLineitem = computeActual("select orderkey from lineitem group by orderkey");
+
+        MaterializedResult actual = computeActual("select p.partkey from part p left join lineitem l on p.partkey = l.orderkey group by p.partkey");
+        assertEquals(actual.getRowCount(), countPart.getRowCount());
+        assertNotEquals(actual.getRowCount(), countLineitem.getRowCount());
+
+        actual = computeActual("select l.orderkey from part p right join lineitem l on p.partkey = l.orderkey group by 1");
+        assertEquals(actual.getRowCount(), countLineitem.getRowCount());
+        assertNotEquals(actual.getRowCount(), countPart.getRowCount());
+
+        actual = computeActual("select p.partkey from part p left join lineitem l on p.partkey = l.orderkey");
+        assertNotEquals(actual.getRowCount(), countLineitem.getRowCount());
+        assertNotEquals(actual.getRowCount(), countPart.getRowCount());
+        assertEquals(actual.getRowCount(), 3500);
+
+        actual = computeActual("select p.partkey from part p right join lineitem l on p.partkey = l.orderkey group by p.partkey");
+        assertNotEquals(actual.getRowCount(), countLineitem.getRowCount());
+        assertNotEquals(actual.getRowCount(), countPart.getRowCount());
+        assertEquals(actual.getRowCount(), 504);
+
+        actual = computeActual("select p.partkey from part p join lineitem l on p.partkey = l.orderkey group by p.partkey");
+        assertNotEquals(actual.getRowCount(), countLineitem.getRowCount());
+        assertNotEquals(actual.getRowCount(), countPart.getRowCount());
+        assertEquals(actual.getRowCount(), 503);
```

Comment:
This should be the same as `SELECT orderkey from lineitem`?

Reply:
no this becomes an inner join for which RemoveUnreferencedJoin rule is not in effect. comment is made in each UT to clarify what is the expected result and what's not